### PR TITLE
[DPE-4683] Typed configuration

### PR DESCRIPTION
**Pull request by dragomirp**
* Bump libs
* Add pydantic vaildation
* Move sockets to accessible directories

Boilerplate for https://github.com/canonical/pgbouncer-operator/pull/343

### CODE REVIEW

File: tests/integration/test_config.py
Comment (dragomirp) [inline tests/integration/test_config.py:1]: Mostly copypaste from PG VM charm

File: tests/integration/relations/pgbouncer_provider/test_data_integrator_ha.py [hunk 1]
```diff
@@ -143,7 +143,7 @@ async def test_remove_tls(ops_test: OpsTest, pgb_charm_jammy):
 @pytest.mark.group(1)
 async def test_remove_vip(ops_test: OpsTest):
     async with ops_test.fast_forward():
-        await ops_test.model.applications[PGB].set_config({"vip": ""})
+        await ops_test.model.applications[PGB].reset_config(["vip"])
```
Comment (dragomirp) [inline tests/integration/relations/pgbouncer_provider/test_data_integrator_ha.py:146]: Empty string will no longer pass the address validation.

File: src/charm.py [hunk 2]
```diff
@@ -129,14 +133,17 @@ def __init__(self, *args):
...
+            self._grafana_agent = COSAgentProvider(
+                self,
+                metrics_endpoints=[
+                    {"path": "/metrics", "port": self.config.metrics_port},
+                ],
+                log_slots=[f"{PGBOUNCER_SNAP_NAME}:logs"],
+                refresh_events=[self.on.config_changed],
+            )
+        except ValueError:
+            logger.warning("Unable to set COS agent, invalid config")
```
Comment (dragomirp) [inline src/charm.py:146]: Data will be validated the first time we access `self.config`

File: src/config.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+
+class CharmConfig(BaseConfigModel):
+    """Manager for the structured configuration."""
+
+    listen_port: PositiveInt
+    metrics_port: PositiveInt
+    vip: Optional[IPvAnyAddress]
+    pool_mode: Literal["session", "transaction", "statement"]
+    max_db_connections: conint(ge=0)
```
Comment (dragomirp) [inline src/config.py:23]: 0 is unlimited db conns.

File: src/relations/db.py [hunk 4]
```diff
@@ -324,12 +324,14 @@ def _on_relation_changed(self, change_event: RelationChangedEvent):
...
-        self.update_connection_info(change_event.relation, self.charm.config["listen_port"])
+        self.update_connection_info(change_event.relation)
 
-    def update_connection_info(self, relation: Relation, port: str = None):
+    def update_connection_info(self, relation: Relation):
         """Updates databag connection information."""
-        if not port:
-            port = self.charm.config["listen_port"]
+        if not self.charm.configuration_check():
+            return
```
Comment (dragomirp) [inline src/relations/db.py:332]: Function will be rerun on config change

File: src/relations/pgbouncer_provider.py [hunk 5]
```diff
@@ -229,7 +227,7 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
 
     def update_connection_info(self, relation):
         """Updates client-facing relation information."""
-        if not self.charm.unit.is_leader():
+        if not self.charm.unit.is_leader() or not self.charm.configuration_check():
```
Comment (dragomirp) [inline src/relations/pgbouncer_provider.py:230]: Function will be rerun on config change.

File: src/charm.py [hunk 6]
```diff
@@ -751,11 +777,14 @@ def render_pgb_config(self, reload_pgbouncer=False, restart=False):
         initial_status = self.unit.status
         self.unit.status = MaintenanceStatus("updating PgBouncer config")
 
+        if not self.configuration_check():
+            return
+
```
Comment (dragomirp) [inline src/charm.py:782]: File will be rerendered on config change.

**Review by marceloneppel — APPROVED** [review]
LGTM!

File: src/charm.py [hunk 7]
```diff
@@ -165,11 +172,14 @@ def create_instance_directories(self):
         """Create configuration directories for pgbouncer instances."""
         pg_user = pwd.getpwnam(PG_USER)
         app_conf_dir = f"{PGB_CONF_DIR}/{self.app.name}"
+        app_run_dir = f"{PGB_RUN_DIR}/{self.app.name}"
 
         # Make a directory for each service to store configs.
         for service_id in self.service_ids:
             os.makedirs(f"{app_conf_dir}/{INSTANCE_DIR}{service_id}", 0o700, exist_ok=True)
             os.chown(f"{app_conf_dir}/{INSTANCE_DIR}{service_id}", pg_user.pw_uid, pg_user.pw_gid)
+            os.makedirs(f"{app_run_dir}/{INSTANCE_DIR}{service_id}", 0o777, exist_ok=True)
```
Comment (taurus-forever) [inline src/charm.py:181]: Should we limit it to read for others? Do they need a write capabilities?

**Pull request merged**